File: ios/SwiftKotlination/SwiftKotlination/Sources/Model/ImageManager.swift
import UIKit.UIImage

protocol ImageManagerProtocol {
    @discardableResult
    func image(with url: String, _ observer: @escaping Observer<UIImage>) -> Disposable
}

struct ImageManager {
    let networkManager: NetworkManagerProtocol
}

// MARK: - Protocol Methods

extension ImageManager: ImageManagerProtocol {
    @discardableResult
    func image(with url: String, _ observer: @escaping Observer<UIImage>) -> Disposable {
        return networkManager
            .observe(.fetchImage(url)) { result in
                switch result {
                case .success(let data):
                    guard let image = UIImage(data: data) else {
                        observer(.failure(NetworkError.invalidResponse))
                        return
                    }

                    observer(.success(image))

                case .failure(let error):
                    observer(.failure(error))
                }
        }
    }
}

File: android/SwiftKotlination/modules/image/src/main/kotlin/fr/jhandguy/image/model/ImageManager.kt
package fr.jhandguy.image.model

import fr.jhandguy.network.model.network.NetworkManagerInterface
import fr.jhandguy.network.model.network.Request
import fr.jhandguy.network.model.observer.Disposable
import fr.jhandguy.network.model.observer.Observer

interface ImageManagerInterface {
    suspend fun image(url: String, observer: Observer<ByteArray>): Disposable
}

class ImageManager(private val networkManager: NetworkManagerInterface) : ImageManagerInterface {
    override suspend fun image(url: String, observer: Observer<ByteArray>): Disposable =
        networkManager.observe(Request.FetchImage(url), observer)
}
